src/components/partnership/partnership.repository.ts

```diff
@@ -259,33 +259,80 @@ export class PartnershipRepository extends DtoRepository(Partnership) {
       .apply(calculateTotalAndPaginateList(Partnership, input));
   }
 
-  async verifyRelationshipEligibility(projectId: ID, partnerId: ID) {
+  async verifyRelationshipEligibility(
+    projectId: ID,
+    partnerId: ID,
+    changeset?: ID
+  ) {
     return (
       (await this.db
         .query()
-        .optionalMatch(node('partner', 'Partner', { id: partnerId }))
-        .optionalMatch(node('project', 'Project', { id: projectId }))
-        .optionalMatch([
-          node('project'),
-          relation('out', '', 'partnership', { active: true }),
-          node('partnership'),
-          relation('out', '', 'partner', { active: true }),
-          node('partner'),
-        ])
+        .subQuery((sub) =>
+          sub
+            .optionalMatch(node('partner', 'Partner', { id: partnerId }))
+            .optionalMatch(node('project', 'Project', { id: projectId }))
+            .optionalMatch([
+              node('project'),
+              relation('out', '', 'partnership', { active: true }),
+              node('partnership'),
+              relation('out', '', 'partner', { active: true }),
+              node('partner'),
+            ])
+            .return(['partner', 'project', 'partnership'])
+            .apply((q) =>
+              changeset
+                ? q
+                    .union()
+                    .match([node('changeset', 'Changeset', { id: changeset })])
+                    .optionalMatch([
+                      node('project'),
+                      relation('out', '', 'partnership', { active: false }),
+                      node('partnership'),
+                      relation('in', '', 'changeset', { active: true }),
+                      node('changeset'),
+                    ])
+                    .optionalMatch([
+                      node('partnership'),
+                      relation('out', '', 'partner', { active: true }),
+                      node('partner'),
+                    ])
+                    .return(['partner', 'project', 'partnership'])
+                : q
+            )
+        )
```

Comment:
The partner and project matched on the left hand side of the union are not considered on the right hand side of the union. Therefore, the current logic matches _any_ project & partner.

Maybe move partner & project out of the subquery & import them on both sides of the union?